File: scripts/lib/atproto/clients/raw.py
from __future__ import annotations

from typing import Any, Optional

from ..transport import TransportLayer


class RawClient:
    """Non-XRPC HTTP methods for admin routes and other raw endpoints."""

    def __init__(self, transport: TransportLayer):
        self._t = transport

    def http_get(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict:
        return self._t.http_get(path, params=params, token=token)

    def http_post(
        self,
        path: str,
        body: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict:
        return self._t.http_post(path, body=body, token=token)

    def xrpc_get(
        self,
        method: str,
        params: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict:
        return self._t.get(method, params=params, token=token)

    def xrpc_post(
        self,
        method: str,
        body: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict[str, Any]:
        return self._t.post(method, body=body, token=token)

    def post_raw(
        self,
        method: str,
        data: bytes,
        content_type: str,
        params: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict[str, Any]:
        """POST raw binary data to an XRPC endpoint."""
        return self._t.post_raw(method, data, content_type=content_type, token=token)

    def xrpc_get_binary(
        self,
        method: str,
        params: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
        headers: Optional[dict[str, str]] = None,
    ) -> tuple[int, str, bytes]:
        return self._t.get_binary(method, params=params, token=token, headers=headers)

File: scripts/lib/deno/clients/raw.ts
import { TransportLayer } from "../transport.ts";

export class RawClient {
  constructor(private transport: TransportLayer) {}

  async httpGet(path: string, params?: Record<string, any>, token?: string) {
    return await this.transport.httpGet(path, params, token);
  }

  async httpPost(path: string, body?: any, token?: string) {
    return await this.transport.httpPost(path, body, token);
  }

  async xrpcGet(method: string, params?: Record<string, any>, token?: string) {
    return await this.transport.get(method, params, token);
  }

  async xrpcPost(method: string, body?: any, token?: string) {
    return await this.transport.post(method, body, token);
  }

  // Aliases for backward compatibility
  async get(method: string, params?: Record<string, any>, token?: string) {
    return await this.xrpcGet(method, params, token);
  }

  async post(method: string, body?: any, token?: string) {
    return await this.xrpcPost(method, body, token);
  }

  async postBinary(method: string, data: Uint8Array, contentType: string, token?: string) {
    return await this.transport.postBinary(method, data, contentType, token);
  }

  async postRaw(
    method: string,
    data: Uint8Array,
    contentType: string,
    options: { token?: string; params?: Record<string, any> } = {}
  ) {
    if (options.params && Object.keys(options.params).length > 0) {
      throw new Error("postRaw does not support params; use xrpcPost for parameterized requests");
    }
    return await this.transport.postBinary(method, data, contentType, options.token);
  }

  async xrpcGetBinary(
    method: string,
    options: { params?: Record<string, any>; token?: string; headers?: Record<string, string> } = {}
  ) {
    return await this.transport.getBinary(method, options.params, options.token, options.headers);
  }
}
